feat: ES2015 features for js inlining (#5567)
JetBrains/kotlin-web-site: docs/topics/js/js-interop.md

## Inline JavaScript

You can inline JavaScript code into your Kotlin code using the [`js()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.js/js.html) function:

```kotlin
fun jsTypeOf(o: Any): String {
    return js("typeof o")
}
```

JavaScript code inlining has full support for [ES2015 features](js-project-setup.md#support-for-es2015-features), including:

* Lambdas ([arrow functions](whatsnew21.md#support-for-generating-es2015-arrow-functions))
* ES classes
* Template strings
* Spread and rest operators
* `const` and `let` variable declarations

> * Generators

Reviewer: Nitpick: Unless there's a hidden logic to the order that I'm not seeing, how about alphabetizing the list?

> Because the parameter of `js` is parsed at compile time and translated to JavaScript code "as-is", it is required to be

Reviewer: ```suggestion
Because the parameter of `js` is parsed at compile time and translated to JavaScript code "as-is", it must be
```

> a string constant. So, the following code is incorrect:

Reviewer: Can you add a little explanation that it's incorrect because the string concatenation turns it into a runtime expression? It took me a couple of minutes to figure this bit out and I think it could help other readers.

Author: Added a comment to the code sample

fun getTypeof() = "typeof"
```

> Instead, for example for the rest operator, use:

Reviewer: ```suggestion
Instead, you can pass strings using the rest operator. For example:
```

Author: I'd focus here more on inlining the mentioned features than passing strings

```kotlin

> fun runSpreadExample() {

Reviewer 2: ```suggestion
fun runSumExample() {
```

Since there is no "spread" anymore - `...nums` is a similar, but another feature called "rest parameters". But in general this impl looks more like it's about sum.

    val sum = js("(...nums) => nums.reduce((a, b) => a + b, 0)")
    println(sum(1, 2, 3, 4))
}
```

> > Invoking `js()` returns a result of type [`dynamic`](dynamic-type.md), which provides no type safety at compile time.

Reviewer: ```suggestion
> Invoking `js()` returns a result of [`dynamic`](dynamic-type.md) type, which doesn't provide type safety at compile time.
```

>
{style="note"}

> ## external modifier

Reviewer: ```suggestion
## `external` modifier
```
Since we have inline formatting in headers now!

fun getTypeof() = "typeof"
```

Instead, for example for the spread operator, use:

```kotlin
fun spreadExample(): dynamic = js("""
    const add = (a, b, c) => a + b + c;

    const nums = [1, 2, 3];
    const sum = add(...nums);

    const a = [1, 2, 3];
    const b = [...a, 4, 5, 6];

    return { sum, b: b };

> """)

Reviewer 2: I feel like this example is too synthetic. For example, the above example for `typeof` actually makes sense to use and it highlights why `js()` in general can be useful. Can we please change this one to a shorter, concise one?

What I mean is something like:

```kotlin
val sum = js("(...nums) => nums.reduce((a, b) => a + b, 0)");
println(sum(1, 2, 3, 4))
```

Of course, it is not filled with everything, but it provides an idea, I suppose.
